### DISCUSSION TITLE
How to understand the disjoint_train_ratio parameter in RandomLinkSplit?

### DISCUSSION THREAD

**Discussion by ki-ljl** [pyg-team/pytorch_geometric#5041] [Q&A] [answered]
When I use ```RandomLinkSplit``` to split the dataset, I get results that I cannot understand:
```python
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
transform = T.Compose([
    T.NormalizeFeatures(),
    T.ToDevice(device),
    T.RandomLinkSplit(num_val=0.1, num_test=0.1, is_undirected=True,
                      add_negative_train_samples=False),
])
dataset = Planetoid('data', name='CiteSeer', transform=transform)
train_data, val_data, test_data = dataset[0]
```
I get the following results:
```python
Data(x=[3327, 3703], edge_index=[2, 7284], y=[3327], train_mask=[3327], val_mask=[3327], test_mask=[3327], edge_label=[3642], edge_label_index=[2, 3642])
Data(x=[3327, 3703], edge_index=[2, 7284], y=[3327], train_mask=[3327], val_mask=[3327], test_mask=[3327], edge_label=[910], edge_label_index=[2, 910])
Data(x=[3327, 3703], edge_index=[2, 8194], y=[3327], train_mask=[3327], val_mask=[3327], test_mask=[3327], edge_label=[910], edge_label_index=[2, 910])
```
In my understanding, ```edge_index=[2, 7284]``` in train_data is ```training supervision edges```, that is, edges used for GNN message passing and aggregation. ```edge_label_index=[2, 3642])``` is the training message edges, that is, the edges used for supervised learning.

Now we try to set ```disjoint_train_ratio``` to 0.5 and we will get the following result:
```python
Data(x=[3327, 3703], edge_index=[2, 3642], y=[3327], train_mask=[3327], val_mask=[3327], test_mask=[3327], edge_label=[1821], edge_label_index=[2, 1821])
Data(x=[3327, 3703], edge_index=[2, 7284], y=[3327], train_mask=[3327], val_mask=[3327], test_mask=[3327], edge_label=[910], edge_label_index=[2, 910])
Data(x=[3327, 3703], edge_index=[2, 8194], y=[3327], train_mask=[3327], val_mask=[3327], test_mask=[3327], edge_label=[910], edge_label_index=[2, 910])
```
It can be found that both ```edge_index``` and ```edge_label_index``` in train_data have become half of the previous ones. The explanation of the ```disjoint_train_ratio``` parameter in the PyG documentation is as follows:
```python
If set to a value greater than 0.0, training edges will not be shared for message passing and supervision. Instead, disjoint_train_ratio edges are used as ground-truth labels for supervision during training. (default: 0.0)
```
Based on the above explanation I am confused by the previous output. 

In my understanding if ```disjoint_train_ratio=0.0``` means ```edge_label_index``` will be included in ```edge_index```, if ```disjoint_train_ratio=0.5``` means ```edge_label_index``` will not be included in ```edge_index``` (and it is), but I am curious why ```edge_index``` is also reduced half?

**Comment by rusty1s** [comment, marked as answer]
IMO, it makes sense that both are reduced by half. Previously, message passing edges and supervision edges were shared, while they are not disjoint in a 50:50 ratio. Does that make sense?

**Reply by ki-ljl** [reply]
Thank you for your prompt reply. Forgive me for another stupid question: is there any meaning to keep the ratio of ```edge_index``` and ```edge_label_index``` in train_data at 2:1? Is it to make the number of the two consistent after negative sampling? Or for other reasons?

**Reply by rusty1s** [reply]
The ratio is 2:1 in your case because you are operating on an undirected graph, that means that bidirectional edges are used for message passing, while only one direction of edges is used for supervision (since one typically uses a symmetric decoder).

**Reply by ki-ljl** [reply]
I see, thank you very much for your reply.

**Reply by accessframework** [reply]
Hi, just to clarify, if we set ```disjoint_train_ration > 0```, ```edge_label_index``` and ```edge_index``` should not share edges. Am I right? However I noticed that it is not the case. If I use,

```
transform = T.RandomLinkSplit(
    num_val=0.1,
    num_test=0.1,
    disjoint_train_ratio=0.3,
    neg_sampling_ratio=1.0,
    add_negative_train_samples=False,
    edge_types=("policy", "conflicts", "policy"),
    rev_edge_types=("policy", "conflicts", "policy")
)

train, val, test = transform(train_data)
``` 

```train```'s ```edge_label_index``` and ```edge_index``` contain common edges. What is happening here? and how can I ensure the absolute disjointness between message passing and supervision edges?

**Comment by sakunaharinda** [comment]
Hi, Can you please explain to me the terms training edges, supervision edges, and message passing edges and how they relate?
